kernels/volk/volk_4ic_deinterleave_8i_x2.h

```diff
@@ -0,0 +1,109 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2020 Derek Kozel
+ *
+ */
+
+/*!
+ * \page volk_4ic_deinterleave_8i_x2
+ *
+ * \b Overview
+ *
+ * Deinterleaves the complex 4-bit integer vector into I & Q vector data
+ * and converts them to 8-bit integers.
+ *
+ * <b> Dispatcher Prototype</b>
+ * \code
+ * void volk_4ic_deinterleave_8i_x2(int8_t* iBuffer, int8t_t* qBuffer, const lv_4sc_t complexVector, unsigned int num_points) \endcode
+ *
+ * \b Inputs
+ * \li complexVector: The complex input vector.
+ * \li num_points: The number of complex data values to be deinterleaved.
+ *
+ * \b Outputs
+ * \li iBuffer: The I buffer output data.
+ * \li qBuffer: The Q buffer output data.
+ *
+ * \b Example
+ * \code
+ * int N = 10000;
+ *
+ * volk_4ic_deinterleave_8i_x2();
+ *
+ * volk_free(x);
+ * \endcode
+ *
+ */
+
+#ifndef INCLUDED_volk_4ic_deinterleave_8i_x2_a_H
+#define INCLUDED_volk_4ic_deinterleave_8i_x2_a_H
+
+#include <inttypes.h>
+#include <stdio.h>
+
+#ifdef LV_HAVE_GENERIC
+
+static inline void volk_4ic_deinterleave_8i_x2_generic(int8_t* iBuffer,
+                                                       int8_t* qBuffer,
+                                                       const int8_t* complexVector,
+                                                       unsigned int num_points)
+{
+    const int8_t* complexVectorPtr = (const int8_t*)complexVector;
+    int8_t* iBufferPtr = iBuffer;
+    int8_t* qBufferPtr = qBuffer;
+    for (unsigned int number = 0; number < num_points; number++) {
+       *iBufferPtr = (*complexVectorPtr) >> 4;
+       iBufferPtr++;
+       *qBufferPtr = ((*complexVectorPtr) << 4) >> 4;
+       qBufferPtr++;
+    }
+}
+
+#endif /* LV_HAVE_GENERIC */
+
+#ifdef LV_HAVE_SSE2
+#include <emmintrin.h>
+
+static inline void volk_4ic_deinterleave_8i_x2_a_sse2(int8_t* iBuffer,
+												   	   int8_t* qBuffer,
+													   const int8_t* complexVector,
+													   unsigned int num_points)
+{
+    // SSE2 algorithm was written by Andrey Semashev, licensed as CC-BY-SA
+    // https://stackoverflow.com/questions/63200053/deinterleve-vector-of-nibbles-using-simd
+	const __m128i mask = _mm_set1_epi32(0x0F0F0F0F);
+	const __m128i signed_max = _mm_set1_epi32(0x07070707);
+
+    unsigned int number = 0;
+    const unsigned int sixteenthPoints = num_points / 16;
+	for (; number < sixteenthPoints; number++)
+	{
+		// Load and deinterleave input half-bytes
+		__m128i input_even = _mm_loadu_si128(((const __m128i*)(complexVector)) + number);
```

Comment:
I suggest to use pointer arithmetic and just increment by `16` after every load.
Also, this kernel is marked as `_a_sse2` (aligned). Therefore, I'd expect aligned loads and stores.